mmdet/core/results/results.py

```diff
@@ -0,0 +1,543 @@
+            v0 = values[0]
+            if isinstance(v0, torch.Tensor):
+                values = torch.cat(values, dim=0)
+            elif isinstance(v0, np.ndarray):
+                values = np.concatenate(values, axis=0)
+            elif isinstance(v0, list):
+                values = list(itertools.chain(*values))
+            elif hasattr(type(v0), 'cat'):
+                values = type(v0).cat(values)
+            else:
+                raise ValueError(
+                    f'Can not concat the {k} which is a {type(v0)}')
+            cat_results[k] = values
+        return cat_results
+
+    def __len__(self):
+        if len(self._results_field):
+            for v in self.results_field.values():
+                return len(v)
+        else:
+            raise AssertionError('This is an empty `InstanceResults`.')
+
+
+class DetectionResults(InstanceResults):
+    """A standard results for detection task, It specify several keys for
+    detection such `bboxes`,`scores`, `labels` and `masks`, etc.
+
+    And overloaded `format_results` can convert these value to a format which
+    is suitable for multi-GPU synchronization
+    """
+
+    def __init__(self, meta=None, num_classes=80, **kwargs):
+        super(DetectionResults, self).__init__(meta=meta)
+        extra_metas = dict(num_classes=num_classes)
+        for k, v in kwargs.items():
+            extra_metas[k] = v
+        self.add_meta_info(extra_metas)
+
+    def new_results(self):
+        """Return a new results with same image meta information and empty
+        results_field."""
+        new_results = self.__class__(num_classes=self.num_classes)
+        new_results.add_meta_info(self.meta_info_field)
+        return new_results
+
+    def format_results(self):
+        results_dict = dict()
+        assert 'scores' in self._results_field
+        assert 'labels' in self._results_field
+
+        assert 'bboxes' in self._results_field or \
+            'masks' in self._results_field, \
+            'DetectionResults at least contains one of ' \
+            '(bboxes, masks) when format the results '
+
+        labels = self.labels.detach().cpu().numpy()
+        if 'bboxes' in self._results_field:
+            # format bboxes results
+            if len(self) == 0:
+                bbox_results = [
+                    np.zeros((0, 5), dtype=np.float32)
+                    for _ in range(self.num_classes)
+                ]
+            else:
+                det_bboxes = torch.cat([self.bboxes, self.scores[:, None]],
+                                       dim=-1)
+                det_bboxes = det_bboxes.detach().cpu().numpy()
+                bbox_results = [
+                    det_bboxes[labels == i, :] for i in range(self.num_classes)
+                ]
+
+            results_dict['bbox_results'] = bbox_results
+        if 'masks' in self._results_field:
+            masks = self.masks.detach().cpu().numpy()
```

Comment:
The scores will not be saved in results_dict when only mask results. Is this correct?

Reply:
This is the original design of mmdet, I also believe it is unreasonable